feat(animation): Implement `AnimationDirection`

SK83RJOSH opened the pull request:
This PR implements `PropertyAnimation.direction` as mentioned in #6260, but not `PropertyAnimation.fill-mode`. Slint's default fill mode could be considered `both`, and as such this PR limits the scope to implementing only that fill mode.

If desired, adding support for `AnimationFillMode` could be done, using the following enum:
```rust
enum AnimationFillMode {
    None = 0,
    Forwards = 1,
    Backwards = 2,
    Both = 3,
}
```

And using bit masks:
```rust
(fill_mode & AnimationFillMode::Forwards) != 0
(fill_mode & AnimationFillMode::Backwards) != 0
```

However, doing so would require fetching the animated property *prior* to it being set by the current state when `None`, `Forwards`, or `Backwards` are chosen. Which may prove tricky, and may not currently be possible. Thus my decision to skip the extra complexity for now and keep this PR as bare bones as possible.

Here are some videos demonstrating the feature:

https://github.com/user-attachments/assets/bbdb6082-c957-4036-94d3-5be79ab50cc8

https://github.com/user-attachments/assets/96f4ae62-8f5c-4d4f-bed5-8beace269923

https://github.com/user-attachments/assets/86d76a43-1f55-42d9-8c07-57c652e09513

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slint-ui/slint?pullRequest=6261) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slint-ui/slint?pullRequest=6261) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slint-ui/slint?pullRequest=6261) it.</sub>

SK83RJOSH commented:
Alrighty, I've tested the lsp, added a test, and updated every bit of related code/documentation I could find.

The only remaining question I have, is whether or not any other code might depend on the old behavior? I noticed that very rarely, widgets revert to `to_value` when the layout is refreshed (by fiddling with other widgets/resizing the window). Not sure if it's simply an existing bug with `from`/`to` states, or something I have caused by breaking assumptions made elsewhere.

I would say it's a non-blocker for merging given the rarity, but I'd love to pinpoint the issue if possible. ❤️

ogoffart commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ All notable changes to this project are documented in this file.
  - The `mod` function was changed to always return a positive value (#6178)
  - Added `AccesssibleRole.list-item` and used it where relevant
  - Added `PointerEventButton.back` and `PointerEventButton.forward` (#6043)
+ - Added `PropertyAnimation.direction` property for controlling animation direction (#6261).
```
Please put Remove the line from the CHANGELOG file and put it in a commit message instead.

ogoffart reviewed (commented):
Thanks a lot for your PR. 🤗

This is very complete with tests and docs, appreciated.

Please rebase to fix the conflict (remove the line in the changelog) and this is good to go.

A test in a .slint file would be extra nice (for example like tests/cases/properties/property_animation.slint) (doesn't need to have all languages, one is enough)

SK83RJOSH commented:
Done! Added tests for all languages already in that file, though took a bit longer than I'd like due to rust compile times and the complexity of the test adapter. Time to see if the CI succeeds. 🙂

ogoffart commented:
> took a bit longer than I'd like due to rust compile times

Yes, the test takes ages to compile; you can use the SLINT_TEST_FILTER env variable to only compile a single test to make it faster
https://github.com/slint-ui/slint/blob/master/docs/testing.md#rust-driver

ogoffart reviewed (approved):
Great!

SK83RJOSH commented on internal/core/properties/properties_animations.rs:
```diff
@@ -71,18 +86,32 @@ impl<T: InterpolatedPropertyValue + Clone> PropertyValueAnimationData<T> {
                 } else {
-                    self.state = AnimationState::Done;
+                    self.state =
+                        AnimationState::Done { interation_count: (current_iteration - 1).max(0) };
```
Whoops, just noticed these are unsigned, so this `max` is useless - I'm pretty sure it's never going to underflow anyway, given we bail out on `iteration_count == 0.`.

ogoffart commented:
Thanks!

tronical commented:
Awesome feature and implementation:)
